FEATURE: Deprecate `triggeringNode` in favour of `site` and `parentNode`

mhsdesign opened the pull request:
Introduce `parentNode` as reference to the parent of the node the template is initially applied on. The parent of the outermost node the Neos UI creates.

**New variables:**

| Variable name                | Type                   | Description                                                     | Availability            |
|------------------------------|------------------------|-----------------------------------------------------------------|-------------------------|
| site                         | `Node`                 | The site node in which the new node be created in               | Global                  |
| parentNode                   | `Node`                 | The node where the new utmost node will be created inside       | Global                  |
| ~triggeringNode~ | `Node`                 | _Deprecated:_ The new node itself which is triggering the template processing | Global                  |


> **Notice**
> `triggeringNode` will be removed with version 3.0

> **Warning**
> The behaviour of `parentNode` changed from version 1.x to version 2.2
> Previously it referenced the parent node of the current template part and its nesting.
> With version 2.0 it was removed and 2.2 reintroduced the variable identifying the parent node of the first/utmost node that will be created.



~We should wait with merging this, as `parentNode` meant something else in V1, and was breaking removed in V2.~

mhsdesign committed:
FEATURE: Deprecated `triggeringNode` in favor of `parentNode`

Introduce `parentNode` as reference to the parent of the node the template is initially applied on. The parent of the outermost node the Neos UI creates.

Depr

Sebobo commented:
So is this PR now breaking or what should we wait for?

mhsdesign committed:
DOCS: `parentNode`

mhsdesign commented:
No its just unexpected, but then again when migrating from `v 1.x` its not less unexpected that `parentNode` will be null, so we might as well give it already new purpose with `2.1`

I added a warning in the readme ;)

> **Warning**
 > The behaviour of `parentNode` changed from version 1.x to 2.x

mhsdesign commented on Classes/Domain/TemplateNodeCreationHandler.php:
```diff
@@ -46,7 +46,9 @@ public function handle(NodeInterface $node, array $data): void
             'data' => $data,
+            // deprectated will be removed in 3.0
             'triggeringNode' => $node,
+            'parentNode' => $node->getParent()
```
@Sebobo and me are still unsure about the naming, as its not clear it always means the same

theilm replied:
We need a materialized node that can be used in flow queries for example. What about  "referenceNode" or "existingNode"? On the other hand, if it really always is the parent node of the new node (formerly known as "triggering node") we provide, why not simply call it that? I know, the behaviour has to change (it's not the parent node of the current node in context any more). However, I'd prefer to have a clear name for the new version, even if it may cause some confusion for someone used to the old version. WDYT?

Sebobo replied:
parentNode still sounds best I agree, but referenceNode would be better documentation wise and would lead to less accidental confusions depending on the context.

mhsdesign replied:
I think referenceNode is ambiguous as one might confuse it with the „references“ feature.

Maybe `contextNode` but I dislike the word context - it’s too overused

Sebobo replied:
Context variables by default cause confusion. I think with `referenceNode` we do the least harm.

mhsdesign replied:
Maybe `insertionReferenceNode` or `anchorNode`


I like anchorNode the best ^^

Sebobo replied:
Nope 😀

mhsdesign replied:
how about `site` for the current siteNode? - should cover the most use cases (like all mine i had ^^)

Sebobo replied:
I think the site is usable but the parent too. Remember, in our customers project they actually want to later adjust things based on where in the tree the template is executed.

mhsdesign replied:
hmm how about just simply `parentSourceNode` and `site`?

mhsdesign replied:
I dont like `referenceNode` or the like because `reference` will let people think of node references

mhsdesign replied:
or for our usecase @Sebobo the current `documentNode` would suffice as well ...

mhsdesign replied:
`triggeringNode` is currently the to be created node but in Neos9 it will  be deprecated and null
`parentNode` will be the parent node independent of the current insertion anchor thingy
`site` will be the site node


Depending on the current node / select node is not something we want to expose as its fragile and depends what the editor did and is nothing predictable.

mhsdesign committed:
FEATURE: Add `site` and `parentSourceNode` variables

mhsdesign committed:
TASK: Rename `parentSourceNode` to `parentNode`

As discussed the "source" doesnt add any information except it tries to minimalize confusion for version 1.x users where node and parentNode meant something else.

mhsdesign commented:
As discussed in todays weekly with christian and paula, we concluded that `parentNode` and `site` are good new variables.
The confusion as `parentNode` existed beforehand with different meaning should be minimal and will be documented.
Version 2 is out since 1 year so enough time has passed for us to make such behaviour change. Also as `parentNode` doesnt exist in 2.0 yet this can be perfectly legal released as minor change.

https://github.com/Flowpack/Flowpack.NodeTemplates/pull/70#discussion_r1648764876
